src/gui/inputdevices/TouchInputHandler.cpp

```diff
@@ -101,8 +117,71 @@ void TouchInputHandler::actionEnd(InputEvent* event)
 {
 	XOJ_CHECK_TYPE(TouchInputHandler);
 
-	this->lastPosX = -1.0;
-	this->lastPosY = -1.0;
+	this->priLastPosX = -1.0;
+	this->priLastPosY = -1.0;
+	this->secLastPosX = -1.0;
+	this->secLastPosX = -1.0;
+	this->lastDiff = 0.0;
+
+	if (this->secondarySequence)
+	{
+		ZoomControl* zoomControl = this->inputContext->getView()->getControl()->getZoomControl();
+		zoomControl->endZoomSequence();
+	}
 }
 
+void TouchInputHandler::zoomStart(InputEvent* event)
+{
+	XOJ_CHECK_TYPE(TouchInputHandler);
+
+	this->secLastPosX = event->absoluteX;
+	this->secLastPosY = event->absoluteY;
+
+	double centerX = (this->priLastPosX + this->secLastPosX) / 2;
+	double centerY = (this->priLastPosY + this->secLastPosY) / 2;
 
+	ZoomControl* zoomControl = this->inputContext->getView()->getControl()->getZoomControl();
+
+	if(zoomControl->isZoomFitMode())
+	{
+		zoomControl->setZoomFitMode(false);
+	}
+
+	zoomControl->startZoomSequence(centerX, centerY);
+}
+
+void TouchInputHandler::zoomMotion(InputEvent* event)
+{
+	XOJ_CHECK_TYPE(TouchInputHandler);
+
+	double sqOldDistance = std::pow(this->priLastPosX - this->secLastPosX, 2) + std::pow(this->priLastPosY- this->secLastPosY, 2);
+
+	if (event->sequence == this->primarySequence)
+	{
+		this->priLastPosX = event->absoluteX;
+		this->priLastPosY = event->absoluteY;
+	} else {
+		this->secLastPosX = event->absoluteX;
+		this->secLastPosY = event->absoluteY;
+	}
+
+	double sqNewDistance = std::pow(this->priLastPosX - this->secLastPosX, 2) + std::pow(this->priLastPosY- this->secLastPosY, 2);
+
+	double diff = sqNewDistance - sqOldDistance;
+
+	if (diff == 0.0)
+	{
+		return;
+	}
+
+	if (this->lastDiff == 0.0)
+	{
+		this->lastDiff = diff;
+		return;
+	}
+
+	ZoomControl* zoomControl = this->inputContext->getView()->getControl()->getZoomControl();
+
+	// TODO: Better algorithm for the zoom value
+	zoomControl->zoomSequenceChange(zoomControl->getZoom() + diff/80000, false);
```

Comment:
This new algortihm should make sure that the same point of the page stays under the finger when zooming. I'm sure you'll find a way to do this!